source/server/connection_handler_impl.cc

```diff
@@ -63,71 +54,140 @@ void ConnectionHandlerImpl::ActiveListener::removeConnection(ActiveConnection& c
+  return listener ? listener->listener_.get() : nullptr;
+}
+
+ConnectionHandlerImpl::ActiveListener*
+ConnectionHandlerImpl::findActiveListenerByAddress(const Network::Address::Instance& address) {
   // This is a linear operation, may need to add a map<address, listener> to improve performance.
   // However, linear performance might be adequate since the number of listeners is small.
   // We do not return stopped listeners.
-  auto listener = std::find_if(
+  auto iter = std::find_if(
       listeners_.begin(), listeners_.end(),
       [&address](const std::pair<Network::Address::InstanceConstSharedPtr, ActiveListenerPtr>& p) {
         return p.second->listener_ != nullptr && p.first->type() == Network::Address::Type::Ip &&
                *(p.first) == address;
       });
 
   // If there is exact address match, return the corresponding listener.
-  if (listener != listeners_.end()) {
-    return listener->second->listener_.get();
+  if (iter != listeners_.end()) {
+    return iter->second.get();
   }
 
   // Otherwise, we need to look for the wild card match, i.e., 0.0.0.0:[address_port].
   // We do not return stopped listeners.
   // TODO(wattli): consolidate with previous search for more efficiency.
-  listener = std::find_if(
+  iter = std::find_if(
       listeners_.begin(), listeners_.end(),
       [&address](const std::pair<Network::Address::InstanceConstSharedPtr, ActiveListenerPtr>& p) {
         return p.second->listener_ != nullptr && p.first->type() == Network::Address::Type::Ip &&
                p.first->ip()->port() == address.ip()->port() && p.first->ip()->isAnyAddress();
       });
-  return (listener != listeners_.end()) ? listener->second->listener_.get() : nullptr;
+  return (iter != listeners_.end()) ? iter->second.get() : nullptr;
+}
+
+void ConnectionHandlerImpl::ActiveSocket::continueFilterChain(bool success) {
+  if (success) {
+    if (iter_ == accept_filters_.end()) {
+      iter_ = accept_filters_.begin();
+    } else {
+      iter_ = std::next(iter_);
+    }
+
+    for (; iter_ != accept_filters_.end(); iter_++) {
+      Network::FilterStatus status = (*iter_)->onAccept(*this);
+      if (status == Network::FilterStatus::StopIteration) {
+        return;
+      }
+      // Check if another listener may have to be used.
+      if (socket_->localAddressReset()) {
+        // Hands off redirected connections (from iptables) to the listener associated with the
+        // original destination address. If there is no listener associated with the original
+        // destination address, the connection is handled by the listener that receives it.
+        ActiveListener* new_listener =
+            listener_->parent_.findActiveListenerByAddress(*socket_->localAddress());
+
+        if (new_listener != nullptr) {
+          // Reset the accepted socket transient state and hand it to the new listener.
+          socket_->clearReset();
+          new_listener->onAccept(std::move(socket_));
+          goto out;
```

Comment:
Please try to restructure the code to avoid goto.

Reply:
I'm working on this, will update the PR today.